### PR TITLE
[extractor/wistia] Add support for channels

### LINKED ISSUE #4748: [wistia]'NoneType' object has no attribute 'decompress'
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting a bug unrelated to a specific site
- [X] I've verified that I'm running yt-dlp version **2022.08.19** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/ytdl-org/youtube-dl#video-url-contains-an-ampersand-and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)

### Provide a description that is worded well enough to be understood

Test link:
https://fast.wistia.net/embed/channel/yvyvu7wjbg?wchannelid=yvyvu7wjbg&wmediaid=1hweyv01o2

Digging further managed to eventually download video from:
https://embed-fastly.wistia.com/deliveries/665ae0923a3b7c9f28abcae4cf73f074827a3389.m3u8

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
./yt-dlp -vU -F "https://fast.wistia.net/embed/channel/yvyvu7wjbg?wchannelid=yvyvu7wjbg&wmediaid=1hweyv01o2"
[debug] Command-line config: ['-vU', '-F', 'https://fast.wistia.net/embed/channel/yvyvu7wjbg?wchannelid=yvyvu7wjbg&wmediaid=1hweyv01o2']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.08.19 [48c88e088] (zip)
[debug] Plugins: ['SamplePluginIE', 'SamplePluginPP']
[debug] Python 3.9.13 (CPython 64bit) - macOS-12.5.1-arm64-arm-64bit 
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: none
[debug] Optional libraries: sqlite3-2.6.0
[debug] Proxy map: {}
WARNING: Falling back to normal extractor since lazy extractor ABCIE does not have attribute _ENABLED; please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U
WARNING: Falling back to normal extractor since lazy extractor ABCIViewIE does not have attribute _ENABLED; please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U

.................. a ton of other warnings like the above but for other extractors...........

[debug] Loaded 1662 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Latest version: 2022.08.19, Current version: 2022.08.19
yt-dlp is up to date (2022.08.19)
[debug] [generic] Extracting URL: https://fast.wistia.net/embed/channel/yvyvu7wjbg?wchannelid=yvyvu7wjbg&wmediaid=1hweyv01o2
[generic] yvyvu7wjbg?wchannelid=yvyvu7wjbg&wmediaid=1hweyv01o2: Downloading webpage
WARNING: [generic] Falling back on generic information extractor
[generic] yvyvu7wjbg?wchannelid=yvyvu7wjbg&wmediaid=1hweyv01o2: Extracting information
[debug] Looking for Brightcove embeds
[debug] Looking for embeds
[debug] Identified a Wistia embed
[debug] [Wistia] Extracting URL: wistia:yvyvu7wjbg
[Wistia] yvyvu7wjbg: Downloading JSON metadata
ERROR: 'NoneType' object has no attribute 'decompress'
Traceback (most recent call last):
  File "./yt-dlp/yt_dlp/YoutubeDL.py", line 1452, in wrapper
    return func(self, *args, **kwargs)
  File "./yt-dlp/yt_dlp/YoutubeDL.py", line 1528, in __extract_info
    ie_result = ie.extract(url)
  File "./yt-dlp/yt_dlp/extractor/common.py", line 670, in extract
    ie_result = self._real_extract(url)
  File "./yt-dlp/yt_dlp/extractor/wistia.py", line 178, in _real_extract
    embed_config = self._download_embed_config('media', video_id, url)
  File "./yt-dlp/yt_dlp/extractor/wistia.py", line 20, in _download_embed_config
    embed_config = self._download_json(
  File "./yt-dlp/yt_dlp/extractor/common.py", line 1030, in download_content
    res = getattr(self, download_handle.__name__)(url_or_request, video_id, **kwargs)
  File "./yt-dlp/yt_dlp/extractor/common.py", line 994, in download_handle
    res = self._download_webpage_handle(
  File "./yt-dlp/yt_dlp/extractor/common.py", line 864, in _download_webpage_handle
    urlh = self._request_webpage(url_or_request, video_id, note, errnote, fatal, data=data, headers=headers, query=query, expected_status=expected_status)
  File "./yt-dlp/yt_dlp/extractor/common.py", line 803, in _request_webpage
    return self._downloader.urlopen(self._create_request(url_or_request, data, headers, query))
  File "./yt-dlp/yt_dlp/YoutubeDL.py", line 3643, in urlopen
    return self._opener.open(req, timeout=self._socket_timeout)
  File "/opt/homebrew/Cellar/python@3.9/3.9.13_2/Frameworks/Python.framework/Versions/3.9/lib/python3.9/urllib/request.py", line 523, in open
    response = meth(req, response)
  File "./yt-dlp/yt_dlp/utils.py", line 1384, in http_response
    io.BytesIO(self.brotli(resp.read())), old_resp.headers, old_resp.url, old_resp.code)
  File "./yt-dlp/yt_dlp/utils.py", line 1321, in brotli
    return brotli.decompress(data)
AttributeError: 'NoneType' object has no attribute 'decompress'
```

### PR DESCRIPTION
**Pull request by coletdjnz**
### Description of your *pull request* and other information

</details>

and popup-video channel urls (with `--no-playlist`)

Closes #4748 
- the underlying issue with urllib crashing when it should return the compressed bytes is fixed in #2861

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [X] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [X] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [X] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [X] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [X] Fix or improvement to an extractor (Make sure to add/update tests)
- [X] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

File: yt_dlp/extractor/wistia.py [hunk 1]
```diff
@@ -145,17 +152,74 @@ class WistiaIE(WistiaBaseIE):
...
-    # https://wistia.com/support/embed-and-share/video-on-your-website
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        embed_config = self._download_embed_config('medias', video_id, url)
+        return self._extract_media(embed_config)
+
+
+class WistiaEmbedIE(WistiaBaseIE):
+    _VALID_URL = False
+    _EMBED_REGEX = [r'<(?:meta[^>]+?content|(?:iframe|script)[^>]+?src)=["\'](?P<url>(?:https?:)?//(?:fast\.)?wistia\.(?:net|com)/embed/(?:iframe|medias)/[a-z0-9]{10})']
```
Comment (pukkandan) [inline yt_dlp/extractor/wistia.py:163]: too long

File: yt_dlp/extractor/wistia.py [hunk 2]
```diff
@@ -145,17 +152,74 @@ class WistiaIE(WistiaBaseIE):
...
     }]
 
-    # https://wistia.com/support/embed-and-share/video-on-your-website
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        embed_config = self._download_embed_config('medias', video_id, url)
+        return self._extract_media(embed_config)
+
+
+class WistiaEmbedIE(WistiaBaseIE):
```
Comment (pukkandan) [inline yt_dlp/extractor/wistia.py:161]: Why does this need to be a separate extractor?

File: yt_dlp/extractor/wistia.py [hunk 3]
```diff
@@ -202,3 +261,74 @@ def _real_extract(self, url):
...
+        channel_id = self._match_id(url)
+        params = parse_qs(url)
+        media_id = traverse_obj(params, ('wmediaid', 0))
+        if media_id:
+            if self.get_param('noplaylist'):
+                self.to_screen(f'Downloading just video {media_id} because of --no-playlist')
+                return self.url_result(f'wistia:{media_id}', 'Wistia')
+            else:
+                self.to_screen(
+                    f'Downloading channel {channel_id}; add --no-playlist to just download video {media_id}')
```
Comment (pukkandan) [inline yt_dlp/extractor/wistia.py:312]: `self._yes_playlist` is made to standardize these messages

File: yt_dlp/extractor/wistia.py (hunk 2 above)
Comment (coletdjnz) [reply yt_dlp/extractor/wistia.py]: I've split it out now.

Some of the embed code was tightly shared between the extractors previously hence I merged it all.

File: yt_dlp/extractor/wistia.py (hunk 1 above)
Comment (coletdjnz) [reply yt_dlp/extractor/wistia.py]: was how it was originally but fixed

**Comment by coletdjnz** [comment]
TODO: https://github.com/yt-dlp/yt-dlp/pull/4990 possibly conflicts with channel embed detection for this, so will need to fix that :thinking:

**Comment by bashonly** [comment]
> TODO: #4990 possibly conflicts/race condition with channel embed detection for this, so will need to fix that thinking

sorry! i didn't mean to step on your PR. by all means revert my changes with this PR. maybe the updated tests will still be useful

**Comment by coletdjnz** [comment]
> > TODO: #4990 possibly conflicts/race condition with channel embed detection for this, so will need to fix that thinking
> 
> sorry! I didn't mean to step on your PR. by all means revert my changes with this PR. maybe the updated tests will still be useful
> 
> if it helps at all, I found that the embedded channel JS parses the query params `wchannelid`, `wgalleryid` and `wseriesid` for ID values in addition to `wmediaid`, `wvideoid` and `wvideo`

its alg. I just need to add a check that it doesn't match a channel I think. good to have for a fallback anyways.

Your findings will be helpful - looks like I missed a few for the `--no-playlist` redirect. 

I wonder where wgalleryid is used :thinking: 

wseriesid probably refers to the individual series within a channel, which I don't differentiate atm.

File: yt_dlp/extractor/wistia.py [hunk 4]
```diff
@@ -7,24 +7,32 @@
...
     try_call,
     try_get,
+    parse_qs,
+    traverse_obj,
+    update_url_query,
 )
 
+import urllib.parse
+import urllib.error
+from base64 import b64decode
```
Comment (pukkandan) [inline yt_dlp/extractor/wistia.py:17]: move to top

File: yt_dlp/extractor/wistia.py [hunk 5]
```diff
@@ -202,3 +250,103 @@ def _real_extract(self, url):
...
+                channel_id, transform_source=lambda x: urllib.parse.unquote_plus(b64decode(x).decode('utf-8')))
+
+        # XXX: the response suggests there can be multiple "series" but I've never seen one
+        series = traverse_obj(data, ('series', 0), default={})
+
+        entries = []
+        for video in traverse_obj(series, ('sections', ..., 'videos', ...)) or []:
+            hashed_id = video.get('hashedId')
+            if hashed_id:
+                entries.append(self.url_result(f'wistia:{hashed_id}', 'Wistia', title=video.get('name')))
```
Comment (pukkandan) [inline yt_dlp/extractor/wistia.py:341]:
```suggestion
        entries = [
        	self.url_result(f'wistia:{video["hashedId"]}', WistiaIE, title=video.get('name'))
        	for video in traverse_obj(series, ('sections', ..., 'videos', ...)) or []
        	if video.get('hashedId')
        ]
```

**Review by pukkandan — APPROVED** [review]
Looks good except for the conflict. You can revert the other commit if it makes sense (ie, if --no-playlist handles it)

**Comment by coletdjnz** [comment]
The url_transparent for the "uploader" was causing issues with metadata (preferring the generic over the actual video metadata).

I've removed it entirely since it doesn't make much sense - the site the video is embedded on is not necessary the uploader?

If we need it then we might need to do some smuggling trickery.

**Comment by pukkandan** [comment]
> the site the video is embedded on is not necessary the uploader?

makes sense. We'll reimplement if anyone actually requests it

**Pull request merged**